docs/design-documents/20231117-better-processors.md

```diff
@@ -0,0 +1,889 @@
+
+We have to differentiate listing actual processors that are attached to actual pipelines
+and listing the specifications of all available processors. We can draw a parallel to
+listing connectors vs. listing connector plugins. We already have a mechanism to list
+actual processors using
+[`ProcessorService.ListProcessors`](https://github.com/ConduitIO/conduit/blob/df048d93e6240c2c3ec5e52af430c05bdfc7a1a3/proto/api/v1/api.proto#L713),
+we need to add a way to list available processors.
+
+Since we are adding pluggable processors we propose to use the same terminology as with
+connectors and introduce **processor plugins**. This would also give us the ability to
+treat them in a similar way and expose them to the user through the plugin service. We
+propose to change the plugin service so that it exposes connector plugins and processor
+plugins on separate endpoints. For backwards compatibility we would still expose connector
+plugins on the same endpoint as before (this would be removed before the v1.0 release).
+
+```protobuf
+service PluginService {
+  // ListPlugins is a DEPRECATED. It returns a list of all connector plugins.
+  rpc ListPlugins(ListPluginsRequest) returns (ListPluginsResponse) {
+    option (google.api.http) = {
+      get: "/v1/plugins"
+      response_body: "plugins"
+    };
+  };
+
+  // ListConnectorPlugins returns a list of all connector plugins.
+  rpc ListConnectorPlugins(ListConnectorPluginsRequest) returns (ListConnectorPluginsResponse) {
+    option (google.api.http) = {
+      get: "/v1/plugins/connectors"
+      response_body: "connectors"
+    };
+  };
+
+  // ListProcessorPlugins returns a list of all processor plugins.
+  rpc ListProcessorPlugins(ListProcessorPluginsRequest) returns (ListProcessorPluginsResponse) {
+    option (google.api.http) = {
+      get: "/v1/plugins/processors"
+      response_body: "processors"
+    };
+  };
+}
+```
+
+Note that builtin processor plugins would be listed in this endpoint alongside
+standalone plugins.
+
+### Pluggable processors and SDK
+
+We want to give Conduit users the ability to provide their own processor in form of a
+plugin. We propose a similar approach to pluggable processors as we already have with
+pluggable connectors. This means that processors would be loaded at startup and added
+to the processor plugin registry. Processors would be addressable using a similar naming
+scheme as connectors (more on
+[referencing connectors](https://conduit.io/docs/connectors/referencing)). Built-in
+processors would be created in a similar manner as built-in connectors, in that they
+would implement the same interface and could essentially be extracted into standalone
+processors (more info on
+[standalone vs built-in connectors](https://conduit.io/docs/connectors/behavior#standalone-vs-built-in-connectors)).
+
+The only difference between pluggable processors and connectors is the proposed plugin
+mechanism. For pluggable processors we are leaning towards using
```

Comment:
Does this mean all processors will be WASM?

Reply:
I assume by "all" you mean "standalone and built-in"? Then no, IMO we should have a similar mechanism as we do with connectors, where they can be run both as built-in or pluggable connectors. Built-in connectors skip the gRPC layer, so similarly built-in processors would skip the WASM layer. Something like this:

![image](https://github.com/ConduitIO/conduit/assets/8320753/bcb490ce-bec1-480a-ac2d-fed09061ff14)

Reply:
@lovromazgon ok but where does the WASM layer live?;-)

Reply:
A part of it is in Conduit (the host), where we expose any utility functions we need and run the WASM binary, the other part of the equation is the Processor SDK which nicely exposes host functions and makes it generally simple to write a processor.